pkg/sync/map.go

```diff
@@ -0,0 +1,171 @@
+package sync
+
+import (
+	"sync"
+
+	"golang.org/x/exp/maps"
+)
+
+// Map is like a Go map[interface{}]interface{} but is safe for concurrent use by multiple goroutines.
+type Map[K comparable, V any] struct {
+	mutex sync.RWMutex
+	data  map[K]V
+}
+
+// NewMap creates map.
+func NewMap[K comparable, V any]() *Map[K, V] {
+	return &Map[K, V]{
+		data: make(map[K]V),
+	}
+}
+
+// Store sets the value for a key.
+func (m *Map[K, V]) Store(key K, value V) {
+	m.StoreWithFunc(key, func() V { return value })
+}
+
+func (m *Map[K, V]) StoreWithFunc(key K, createFunc func() V) {
+	v := createFunc()
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	m.data[key] = v
+}
+
+// Load returns the value stored in the map for a key, or nil if no value is present. The loaded value is read-only and should not be modified.
+// The ok result indicates whether value was found in the map.
+func (m *Map[K, V]) Load(key K) (value V, ok bool) {
+	return m.LoadWithFunc(key, nil)
+}
+
+func (m *Map[K, V]) LoadWithFunc(key K, onLoadFunc func(value V) V) (V, bool) {
+	m.mutex.RLock()
+	value, ok := m.data[key]
+	defer m.mutex.RUnlock()
+	if ok && onLoadFunc != nil {
+		value = onLoadFunc(value)
+	}
+	return value, ok
+}
+
+// LoadOrStore returns the existing value for the key if present. The loaded value is read-only and should not be modified.
+// Otherwise, it stores and returns the given value. The loaded result is true if the value was loaded, false if stored.
+func (m *Map[K, V]) LoadOrStore(key K, value V) (actual V, loaded bool) {
+	return m.LoadOrStoreWithFunc(key, nil, func() V { return value })
+}
+
+func (m *Map[K, V]) LoadOrStoreWithFunc(key K, onLoadFunc func(value V) V, createFunc func() V) (actual V, loaded bool) {
+	m.mutex.RLock()
+	v, ok := m.data[key]
+	m.mutex.RUnlock()
+	if ok {
+		if onLoadFunc != nil {
+			v = onLoadFunc(v)
+		}
+		return v, ok
+	}
+	v = createFunc()
+	m.mutex.Lock()
+	m.data[key] = v
+	m.mutex.Unlock()
+	return v, ok
```

Comment:
this need to be guard with `m.mutex.Lock()` otherwise some goroutine can use create object at same time.

Reply:
What do you mean - even if I lock this whole method the lock will be unlocked when method ends. So multiple goroutines will still access the object concurrently, but that should be handled by the object itself.